### PR TITLE
[api] Allow to disable registration API #98

### LINKED ISSUE #98: [feature] Allow to disable the register API endpoint
### Premise

Before working on this, please see how similar features were implemented:

- https://github.com/openwisp/openwisp-radius/issues/96
- https://github.com/openwisp/openwisp-radius/issues/160
- https://github.com/openwisp/openwisp-radius/issues/219

For similar features I mean:

- having a global setting in settings.py
- having a setting for each org in the database, editable via admin interface, this field defaults to NULL and falls back to the global default unless explicitly modified by the org admins 

### Use case

There may be cases in which some or all the organizations using the system may not want to allow users to register accounts.

### 1 step: add a global setting

Add a new setting in settings.py to control whether the registration API is enabled or not at a global level.

The setting will default to True (enabled).

Make sure to document this setting.

### 2 step: OrganizationRadiusSettings

Add a boolean field which indicates whether the registration API is enabled or not.
The boolean can have one of the following values:

- NULL: this is the default
- true: enabled
- false: disabled

Make sure the API view code checks this, if the API endpoint is disabled the API shall return 403.

This needs automated tests.

### Docs

Need to be documented:

- the settings must be listed in the settings page but also mentioned and linked from the API page
- the OrganizationRadiusSettings can be documented in the API page (add a screenshot please)

### PR DESCRIPTION
**Pull request by Purhan**
Closes #98

### CODE REVIEW

**Comment by Purhan** [comment]
@nemesisdesign I've fixed the URL syntax but I really have no idea why CI is failing. Migration checks passed locally :confused:

File: tests/openwisp2/sample_users/migrations/0004_auto_20210118_1211.py
Comment (nemesisdesign) [inline tests/openwisp2/sample_users/migrations/0004_auto_20210118_1211.py:27]: are you sure this migration file is necessary? It shouldn't be.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: tests/openwisp2/sample_users/migrations/0004_auto_20210118_1211.py
Comment (Purhan) [reply tests/openwisp2/sample_users/migrations/0004_auto_20210118_1211.py]: Thanks,
yes this should not be added. I don't understand why it was generated in first place.

**Review by nemesisdesign — COMMENTED** [review]
#220 and #222 will have impact on this, please take a look at those PRs.
Let's wait for #222 to get merged and then rebase this PR and change the numbers of the migration files.
Please ensure consistency.

**Comment by nemesisdesign** [comment]
> #220 and #222 will have impact on this, please take a look at those PRs.
> Let's wait for #222 to get merged and then rebase this PR and change the numbers of the migration files.
> Please ensure consistency.

PRs have been merged, please @Purhan can you rebase on the lastest master, fix conflicts?
Again I want to stress to take a look at the changes introduced in #220 / #222 to ensure consistency (especially with the admin fields and model validation).

**Comment by Purhan** [comment]
@nemesisdesign this is ready for review

File: openwisp_radius/admin.py
Comment (nemesisdesign) [inline openwisp_radius/admin.py:564]: Let's just call this `registration_enabled`.
Ensure this field is shown before `first_name`.

File: openwisp_radius/api/permissions.py
Comment (nemesisdesign) [inline openwisp_radius/api/permissions.py:38]: just do: `view.organization.radius_settings.registration_enabled` for consistency with the other views

File: openwisp_radius/api/permissions.py
Comment (nemesisdesign) [inline openwisp_radius/api/permissions.py:8]: use `from . import settings as app_settings` for consistency, use `app_settings.REGISTRATION_API_ENABLED` in the code, which also makes it easier to mock/change the value during tests.

File: openwisp_radius/admin.py
Comment (nemesisdesign) [inline openwisp_radius/admin.py:568]: make default translatable please

File: openwisp_radius/admin.py
Comment (nemesisdesign) [inline openwisp_radius/admin.py:569]: let's make it more user friendly: `Enabled` (flag as translatable)

File: openwisp_radius/admin.py
Comment (nemesisdesign) [inline openwisp_radius/admin.py:570]: let's make it more user friendly: `Disabled` (flag as translatable)

File: openwisp_radius/base/models.py
Comment (nemesisdesign) [inline openwisp_radius/base/models.py:1092]: looks like `_GET_MOBILE_PREFIX_HELP_TEXT` should be changed to `_GET_REGISTRATION_ENDPOINT_HELP_TEXT`

File: openwisp_radius/base/models.py
Comment (nemesisdesign) [inline openwisp_radius/base/models.py:1091]: let's sinmplify this to `registration_enabled`

File: openwisp_radius/base/models.py
Comment (nemesisdesign) [inline openwisp_radius/base/models.py:155]: I didn't notice these messages being prefixed with _GET before, it's a bit confusing to me since _get is usually used as a prefix for functions, not constants.

Anyway, let's just change this one: `_REGISTRATION_ENABLED_HELP_TEXT`.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Good progress @Purhan.

While testing it manually I found this situation which looks like is not working as intended.
Given the default value which is `True`, In the admin, when I set the value

- to `True`, it writes `True` in the DB
- to `False`, it writes `NULL` in the DB and it goes back to being `Default (True)`

It should work in the opposite way: when the value equals the default, NULL (`None`) shall be written.

**Comment by Purhan** [comment]
@nemesisdesign Done, I have implemented all the requested changes (https://github.com/openwisp/openwisp-radius/pull/221/commits/1257847a22cef0b81c25cc70da7448d5d46079af).

> Good progress @Purhan.
> 
> While testing it manually I found this situation which looks like is not working as intended.
> Given the default value which is `True`, In the admin, when I set the value
> 
>     * to `True`, it writes `True` in the DB
> 
>     * to `False`, it writes `NULL` in the DB and it goes back to being `Default (True)`
> 
> 
> It should work in the opposite way: when the value equals the default, NULL (`None`) shall be written.

I fixed this in https://github.com/openwisp/openwisp-radius/pull/221/commits/757c1a0d4e71dec619deb7b8551f0e76a008d709 (please ignore the additional changes in the same commit)

**Review by nemesisdesign — APPROVED** [review]
Thanks @Purhan! I will merge in `dev` to do some minor improvements before merging to `master`, next time please push to a branch in this repository directly instead of using your fork.

**Pull request merged**
